src/audio/google_audio_post_processing.c

```diff
@@ -0,0 +1,497 @@
+{
+	/* Check version from ABI header */
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
+		comp_err(dev, "gapp_ctrl_set_data(): invalid version");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return gapp_ctrl_set_bin_data(dev, cdata);
+	default:
+		comp_err(dev, "gapp_ctrl_set_data(): invalid cdata->cmd");
+		return -EINVAL;
+	}
+}
+
+static int gapp_get_comp_config(struct comp_data *cd, struct sof_ipc_ctrl_data *cdata,
+				int max_data_size)
+{
+	int ret;
+
+	if (sizeof(cd->config) > max_data_size)
+		return -EINVAL;
+
+	ret = memcpy_s(cdata->data->data, max_data_size, cd->config, sizeof(cd->config));
+	if (ret)
+		return ret;
+
+	cdata->data->abi = SOF_ABI_VERSION;
+	cdata->data->size = sizeof(cd->config);
+	return 0;
+}
+
+static int gapp_get_internal_config(struct comp_data *cd, struct sof_ipc_ctrl_data *cdata,
+				    int max_data_size)
+{
+	int data_size;
+	int blob_size = cdata->num_elems;
+
+	if (blob_size > max_data_size)
+		return -EINVAL;
+
+	data_size =
+		GoogleAudioPostProcessingGetConfig(cd->state, cdata->data->type, cdata->msg_index,
+						   (uint8_t *)cdata->data->data, blob_size);
+
+	if (data_size < 0)
+		return -EINVAL;
+
+	cdata->data->abi = SOF_ABI_VERSION;
+	cdata->data->size = data_size;
+
+	return 0;
+}
+
+static int gapp_ctrl_get_bin_data(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata,
+				  int max_data_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (!cd)
+		return -ENODEV;
+
+	comp_dbg(dev, "gapp_ctrl_get_bin_data(): type = %u, index = %u, size = %d",
+		 cdata->data->type, cdata->msg_index, cdata->num_elems);
+
+	switch (cdata->data->type) {
+	case SOF_GAPP_CONFIG:
+		return gapp_get_comp_config(cd, cdata, max_data_size);
+	case SOF_GAPP_MODEL:
+		COMPILER_FALLTHROUGH
+	case SOF_GAPP_DATA:
+		return gapp_get_internal_config(cd, cdata, max_data_size);
+	default:
+		comp_err(dev, "gapp_ctrl_get_bin_data(): unknown binary data type");
+		return -EINVAL;
+	}
+}
+
+static int gapp_ctrl_get_data(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata,
+			      int max_data_size)
+{
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return gapp_ctrl_get_bin_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "smart_amp_ctrl_get_data(): invalid cdata->cmd");
+		return -EINVAL;
+	}
+}
+
+static int gapp_cmd(struct comp_dev *dev, int cmd, void *data, int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	switch (cmd) {
+	case COMP_CMD_SET_VALUE:
+		return gapp_ctrl_set_val(dev, cdata);
+	case COMP_CMD_GET_VALUE:
+		return gapp_ctrl_get_val(dev, cdata, max_data_size);
+	case COMP_CMD_SET_DATA:
+		return gapp_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return gapp_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "gapp_cmd(): unhandled command %d", cmd);
+		return -EINVAL;
+	}
+}
+
+static struct comp_dev *gapp_new(const struct comp_driver *drv, struct sof_ipc_comp *comp)
```

Comment:
comp_new signature and the way ipc config being passed in have changed since tgl-013.
#include <sof/audio/ipc-config.h>
...
static struct comp_dev *gapp_new(const struct comp_driver *drv, struct comp_ipc_config *config,void *spec) {
struct ipc_config_process *ipc_gapp = spec;
dev = comp_alloc(drv, sizeof(*dev));
dev->ipc_config = *config;
...
}